### PR TITLE
Add .env support

### LINKED ISSUE #2375: Add .env support, take 2
Setting the `FLASK_APP` and `FLASK_DEBUG` for every new terminal session is inconvenient. The CLI could load a `.env` file that contained these values instead. Since Heroku is a popular deployment option, this would also help developers who want the same scheme locally.

Instead of writing our own parser, we should use [python-dotenv](https://github.com/theskumar/python-dotenv). This reduces maintenance burden on us and can be an optional dependency. If it is installed, the CLI will load the `.env`, otherwise it will skip it.

Previously discussed in #2164.

---

@miguelgrinberg raised a good point in the previous thread. `.env` is typically used to store sensitive configuration. The convenience of setting `FLASK_APP=` there could encourage users to commit the file publicly, which would be bad.

A solution would be to load two files, `.flask-env` and `.env`, and prominently document that `.flask-env` can be committed but should not contain secrets.

### PR DESCRIPTION
**Pull request by davidism**
If installed, [python-dotenv](https://github.com/theskumar/python-dotenv#readme) is used to load environment variables from `.env` and `.flaskenv` files. Manual vars override `.env`, which overrides `.flaskenv`. `.flaskenv` is intended for public vars, `.env` is intended for private vars that won't be committed to the repository.

The working directory is set to the location of the files, with the assumption that they are in the top level of the project, to facilitate local package imports (#2383).

The files are only loaded when calling the `flask` command or `app.run`, and can be disabled by passing `load_dotenv=False` to a custom `FlaskGroup` or `app.run`. If a user wants to load dotenv files in production, they can call `flask.cli.load_dotenv` manually from a `wsgi.py` file.

Closes #2375
Based off #2164 

---

Example, assuming python-dotenv is installed:

`flaskr/app.py`:

```python
from flask import Flask

def create_app():
    return Flask('flaskr')
```

`.flaskenv`:

```
FLASK_APP=flaskr.app
```

```
flask run
# no need to set FLASK_APP
# no need for intermediate file to call create_app
```

### CODE REVIEW

**Comment by mitsuhiko** [comment]
Looks good. I feel like we should add feature requires to setup.py so you can install `Flask[signals]` or `Flask[dotenv]`.

**Comment by miguelgrinberg** [comment]
In my opinion, loading the .env file only when using `flask run` or `app.run()` can have unexpected results that may take people a long time to debug. The `.flaskenv` file feels like something that makes sense only when using the development server, but `.env` feels like something you'll use regardless of the server. What do you think about always importing `.env`, so that you also get it when running gunicorn/uwsgi/etc.?

**Comment by mitsuhiko** [comment]
@miguelgrinberg i agree to some degree but at the same time a production deployment of flask most likely has a completely untrustworthy CWD and might even be in site-packages. We had similar issues with the instance folder few people are using.

Not sure if a .env file in prod would work reliably for users even.

**Comment by miguelgrinberg** [comment]
@mitsuhiko Yeah, I can see the unpredictable CWD in production being a problem. Just noticed that @davidism added a note in the docs recommending calling `load_dotenv()` manually in production. That is a good compromise.

File: flask/cli.py [hunk 1]
```diff
@@ -472,12 +487,60 @@ def list_commands(self, ctx):
...
         if obj is None:
             obj = ScriptInfo(create_app=self.create_app)
+
         kwargs['obj'] = obj
         kwargs.setdefault('auto_envvar_prefix', 'FLASK')
-        return AppGroup.main(self, *args, **kwargs)
+        return super(FlaskGroup, self).main(*args, **kwargs)
+
+
+def load_dotenv():
```
Comment (miguelgrinberg) [inline flask/cli.py]: I would add target directory where to look for the dot files as an optional argument.

File: flask/cli.py [hunk 2]
```diff
@@ -539,7 +561,9 @@ def load_dotenv():
 
         dotenv.load_dotenv(path)
```
Comment (miguelgrinberg) [inline flask/cli.py]: I think before calling `dotenv.load_dotenv()` here you need to make sure the path found by dotenv did not go above `top_dir`. The `top_dir` does not appear to be enforced as it is. For example, with a `.env` file in examples/flaskr and `top_dir=examples/flaskr/flaskr`, the `.env` file is imported anyway (note that I added a print right above `dotenv.load_dotenv()` to help me test this):

```
>>> from flask.cli import load_dotenv
>>> load_dotenv(top_dir='/home/miguel/flask/examples/flaskr/flaskr')
 * Loading environment variables from "/home/miguel/flask/examples/flaskr/.env"
```

File: flask/cli.py [hunk 3]
```diff
@@ -520,15 +520,37 @@ def load_dotenv():
...
+        if not os.path.isabs(top_dir):
+            raise ValueError("'top_dir' must be an absolute path.")
+
+        cwd = os.path.realpath(os.getcwd())
+        top_dir = os.path.realpath(top_dir)
+
+        # If the top dir is below the current dir, store the current dir and
+        # change to the top dir temporarily. This disables setting the dir
+        # based on found env files.
+        if os.path.commonprefix((cwd, top_dir)) == cwd:
```
Comment (miguelgrinberg) [inline flask/cli.py]: What happens if `top_dir` is not below `cwd`? If the reverse is true (`cwd` below `top_dir`) then you can leave the cwd alone and let dotenv do its lookup, but if `cwd` and `top_dir` are in different branches, I think you still want to change the cwd to `top_dir`. This would be the use case of having an unrelated cwd, and still wanting to import your settings from the .env file located with the app.

**Comment by davidism** [comment]
I was a little hasty pushing the args commit, need to fix some of the logic and add a test.

File: flask/cli.py (hunk 3 above)
Comment (davidism) [inline flask/cli.py]: This is backwards right now. What I intended was `common_prefix != top`, to find when the top dir is not an ancestor of the current dir. Also need to check if the prefix plus the rest of top dir is actually a director, like 3.6's `commonpath`.

When `top_dir` is not an ancestor, I don't want to permanently change it for the same reason mitsuhiko commented above: we don't know the relationship between the working dir and the top dir. For example, if the package is installed, the `.env` file might be in `/etc/myapp/config`, which probably isn't the right cwd either. Supervisor, uWSGI, and other runners have the ability to set the working dir if desired.

Basically, if the user specifies a non-ancestor top dir, I'm assuming they will be the ones managing the working directory.

File: flask/cli.py (hunk 2 above)
Comment (davidism) [inline flask/cli.py]: Yeah, I forgot to make it actually stop at `top_dir`.

**Comment by davidism** [comment]
Moving comment about `top_dir` behavior out of review:

When `top_dir` is not an ancestor, I don't want to permanently change to it for the same reason mitsuhiko commented above: we don't know the relationship between the working dir and the top dir. For example, if the package is installed, the `.env` file might be in `/etc/myapp/config`, which probably isn't the right cwd either. Supervisor, uWSGI, and other runners have the ability to set the working dir if desired.

Basically, if the user specifies a non-ancestor top dir, I'm assuming they will be the ones managing the working directory.

**Comment by miguelgrinberg** [comment]
With this change as it is, I really have no good reason to stop using python-dotenv directly, as that gives me a better experience that is consistent across dev/test/prod configurations. What I was thinking this would look like is something like this:

    def load_dotenv(path=None):
        # ...

If `path` is not passed, then you call `find_dotenv()` to locate the `.env` and/or `.flaskenv` files (not sure `top_dir` is worth the trouble). But then I can also do this:

    load_dotenv(join(dirname(__file__), '.env')

And this loads the .env directly from the place I know I have it. In this second option there would be no need to look for `.env` and `.flaskenv`, or to search parent directories, you will just go and load from the filename I give you, regardless of its name and/or location. For those developers that want to use `.env` just for development, they can just let Flask look for these files, and for those that want to also load an explicit .env file, they can do it in a way that feels a bit more integrated.

Also let's not forget that the `flask` command is sometimes also used in production. As an example, with the current changes, both Flask-Migrate and Flask-Alembic's commands would be used in production, and would be importing .env and .flaskenv files, while the main server running through gunicorn/uwsgi/etc. would not, and this seems like an inconsistency to me.

**Comment by davidism** [comment]
I think your suggestion to just have a `path` argument is good, but at that point I think we should just tell users to call `dotenv.load_dotenv` themselves, they don't get anything from calling `cli.load_dotenv` in that case. I'll push a rollback for most of the previous commit.

---

To give some context of how I envision using this, here's what my project setup looks like in production:

```
/home/project/  # owned by "project" user
    venv/  # added to path so it doesn't need to be activated
    .flaskenv
    uwsgi.ini
    wsgi.py  # optional, usually not needed for my setups
```

A build of the project is installed with `pip install -U project.whl uwsgi`. uWSGI is run from a systemd unit, sets the user and working directory to "project". Units have a directive `EnvironmentFile`, so we don't even need to load the `.env` file with Python. Pretty much any runner can set the working directory, so even if it can't load an env file, calling `load_dotenv()` from a `wsgi.py` file will do the right thing at that point.

Meanwhile, for the `flask` command: SSH into the machine as the "project" user. You're already in the home directory containing the `.flaskenv` file, so run `flask shell` and it's loaded as expected.

**Comment by quanpower** [comment]
When i use dotenv in Flask 0.1.2.2, i set a FLASK_APP in the .env or .flaskenv, but it also can't find the FLASK_APP when i run 'flask run'.Not suppoted in 0.12.2?

**Comment by ThiefMaster** [comment]
@quanpower as you can see in the changelog, it's not part of 0.12.
